### ISSUE TITLE
fix up handling of bvalue scaling and bvec normalisation

### ISSUE THREAD

**Issue by jdtournier** [MRtrix3/mrtrix3#1487]
OK, prompted by Chris Rorden, based on [this discussion](https://github.com/rordenlab/dcm2niix/issues/245), I've had a look into our handling of the bvalue scaling, and it's fair to say that it's a bit messy at the moment... 

To start off with, following #60, the DICOM import code and any subsequent conversion will leave the DW scheme unmodified: no normalisation of bvecs, no bvalue scaling. All those modifications happens in the app processing the data. 

Where this is problematic is for 3rd party apps that rely on `mrconvert` for the DICOM to NIfTI conversion: if they expect the information to be usable as-is, they might come unstuck on e.g. multi-shell data acquired on a Siemens.

The other situation where this is annoying is when you have data where the bvalue scaling _shouldn't_ happen: you then need to remember to pass the `-bvalue_scaling false` option every time you use a command that uses the DW encoding. There's no simple way to store the information as you expect it to be used, rather than as the scanner reported it.

So that argues for at least providing options to control that within `mrconvert`. We may also need control over normalisation of the bvecs without bvalue scaling (for instance, I just came across some Bruker data with non-unit vectors, but correct bvalues). At least that would provide users with some control over the interpretation of the DW table that they can apply once at DICOM import and then not have to worry about it, in those cases where it would otherwise be annoying.

But equally pressing is to sort out the glaring inconsistencies in how we handle these things currently. This is what I've observed from a multi-shell dataset that needs bvalue scaling:

---

```
mrinfo DICOM/ -dwgrad
```
shows the correct encoding (scaled by b-value), as it should.

---
```
mrconvert DICOM/ dwi.nii -export_grad_fsl bvecs bvals
mrconvert DICOM/ dwi.nii -export_grad_mrtrix enc.b
```
results in `bvecs`/`bvals` or an `enc.b` with no scaling by b-value, no unit normalisation of gradient vectors.

---

```
mrinfo DICOM/ -export_grad_fsl bvecs bvals
mrinfo DICOM/ -export_grad_mrtrix enc.b
```
results in `bvecs`/`bvals` or an `enc.b` _with_ scaling by b-value, and unit normalised vectors. 

---

```
mrinfo dwi.nii -fslgrad bvecs bvals -dwgrad
mrinfo dwi.nii -grad enc.b -dwgrad
```
both report whatever was stored in the corresponding `bvecs`/`bvals` or `enc.b`: no bvalue scaling, no unit normalisation. Yet this is supposed to report the DW encoding as it would be interpreted in our apps - which it gets right if run on the DICOM directly (see the first call above). I'm not even sure how that's possible, there must be some weird code path in there...

---

I'm not sure how to deal with this properly, but open to suggestions. I think at the very least, we need to ensure consistency in how the `-export_grad_*` options work, and in how the gradient table is interpreted, regardless of where it came from (DICOM headers or provided by `-grad` or `-fslgrad`). I also reckon we should add options to control the bvalue scaling and bvecs normalisation separately. What I'm not sure about is whether these should be provided as part of the generic DW gradient table import (or export?) option group, and whether they should be performed by default (which I'm a little reluctant to do, there was a reason we went along with #60). Thoughts welcome.

**Comment by bjeurissen** [comment]
This is the kind of trouble that is similar to what I have highlighted in #980. I am in favor of dealing with this _once_  in ``mrconvert`` (with appropriate defaults and options to override), and then have all the subsequent commands rely on a well specified and consistent ``dw_scheme``. Having each command deal with this is tedious for us and confusing and easy to mess up for users. If a power user still wants to see the original scheme, we should just keep it as a separate header entry in the ``mrconvert`` output. I am perfectly fine with the fact that this means that subsequent exports to other formats will lose the untouched ``dw_scheme``. I think this is exactly what most users expect: a dw_scheme or bvals/bvecs combination that will work.

**Comment by thijsdhollander** [comment]
Not 100% sure I've _fully_ understood the issue at hand (but I think I get it); but if I do, then I'm agreeing with @bjeurissen's general proposed strategy here:

> I am in favor of dealing with this once in mrconvert (with appropriate defaults and options to override), and then have all the subsequent commands rely on a well specified and consistent dw_scheme.

I'm personally also fully ok with with the fact that

>  ... this means that subsequent exports to other formats will lose the untouched dw_scheme. I think this is exactly what most users expect: a dw_scheme or bvals/bvecs combination that will work.

The primary goal here should probably be to avoid that "most" users would make, unknowingly, important mistakes.  ...and even if something went wrong, it can then in most cases be traced back to a well identified point in their pipeline: the importing of their data (i.e. `mrconvert`).  That'll make support much easier, and less confusing both for us as well as those users.

So well, :+1:  from me for @bjeurissen's opinion on this one.

**Comment by jdtournier** [comment]
Yes, I'm also rapidly coming to similar conclusions. The problem though is how to handle externally-supplied encodings, either via `-grad` or `-fslgrad`, where we can't guarantee the correctness of the information. I think what we want is all the options for normalisation / rescaling in the GradImport options, but also provide the corrected gradient table in any output produced (either in output image header, or via the `-export_grad_*` options), rather than the original (which is the current default, and the point of  #60 originally). So this would essentially revert #60, but add further control on the steps performed at import. Does that sound OK? 

Also, how do we handle a gradient table if it's in the header? Presumably by that stage, it's already been fixed up and it's good to go as-is, so we don't want any of the gradient import options to apply. But how do we handle the `-grad` option? Do we by default assume it's in the correct format already, and also leave it untouched _unless_ additional GradImport options are provided? What about bvecs/bvals supplied via `-fslgrad`? Do we also treat them as correct, and leave them unmodified unless modifications are explicitly requested? So this would mean that the DICOM import would perform different actions by default than the others? 

Or do we assume that we should perform vector normalisation and b-value scaling by default every time? I reckon if the scheme has already been corrected, then these actions would leave everything as-is (although there is scope for very minor precision errors to creep in). So maybe the safest is to by default always apply both actions when importing from external sources (whether DICOM, or `-grad`/`-fslgrad`), but provide an option to disable these modifications. We would also leave any information found in the mif header unmodified...? Not sure about that last one, it might be tricky to achieve given how the code is organised.

Maybe another option would be to by default unit-normalise the vectors, but first check whether any of the vectors norms differ significantly from zero, and if they do, trigger a b-value rescale. This would avoid precision errors creeping into the b-values, I reckon. And also provide an opportunity to warn the user that the b-value rescaling is being applied, which might not be a bad thing.

What do you all reckon?

**Comment by Lestropie** [comment]
Check `dwipreproc` operation with `bvecs` / `bvals`: I'm thinking it could be provided with multi-shell data where, if scaled correctly, there is insufficient volumes to be considered as shells, but if erroneously not scaled, the number of volumes in the single shell is sufficient. If `eddy` actually succeeds, that would suggest that it's not itself been scaling b-values, and therefore internally the GP model may have been set up incorrectly.

**Comment by jdtournier** [comment]
OK, regarding `dwipreproc`: as we thought, I think there is a problem with the current handling of the b-value scaling:

In the following, `dwi.mif` is a dataset with a DW scheme that requires b-value scaling, as you can verify here:
```
$ mrinfo dwi.mif -shell_bvalues
46.1538 400 900 
$ mrinfo dwi.mif -shell_bvalues -bvalue_scaling 0
0 900 
```
If I convert that to NIfTI, exporting the DW scheme to bvecs/bvals, no b-value scaling is performed:
```
$ mrconvert dwi.mif dwi.nii -export_grad_fsl bvecs bvals
mrconvert: [100%] copying from "dwi.mif" to "dwi.nii"
$ cat bvals 
0 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900 900
```
So at that point, we're relying on `eddy` interpreting the bvecs/bvals as we do... We may have to do what @Lestropie suggests to see whether `eddy` deals with this.

**Comment by jdtournier** [comment]
Had a good think about this today, just jotting down a few thoughts: 

- There's two ways of storing the DW scheme in the `Header`: as a full-blown Eigen::Matrix, or as the text entries that will eventually be committed to the `mif` header. 
  - Currently, we stick to the text version, since it's then relatively easy to just treat the whole things in the same way as regular entries: they get copied over as-is to newly-created Headers, etc. Makes things relatively simple to manage from that point of view. 
  - But if we want to ensure that the DW scheme is always valid as soon as the image has been read, this requires parsing - so converting from text to floating-point and then back feels a bit pointless - especially if the application then requests the DW scheme for processing, at which point it needs to be converted back to floating-point...
  - Based on the above, my preference would be to store the DW scheme in a `Eigen::MatrixXd`, which would be a member of `Header` (this is actually how things used to be back in the day). In this case, we'd need to make sure the Header `keyval` table is cleared of any stray `dw_scheme` entries to avoid issues. 

- We can introduce all the modifications and checks within the `Header` `open()`, `create()` & `scratch()` calls:
  - for `open()`: verify that the scheme matches the number of volumes, unit-normalise and (optionally) b-value scale. The scheme should then be ready to use. There is a question mark as to how to handle invalid schemes (i.e. that don't match the number of volumes). The sensible option here I guess would be to discard the scheme and issue a warning about it, but I'm not sure. 
  - for `create()` & `scratch()`: verify that the scheme still matches the number of volumes, and if not, stash it in the `prior_dw_scheme` entries. 

- There's a fundamental problem with how to handle externally-supplied gradient tables: we shouldn't process the corresponding command-line options automatically within the `Header::open()` call, since there will be cases where multiple input images are provided, and the backend won't have knowledge of which image that scheme belongs to. 
  - This is currently not a problem since the application itself needs to request the DW scheme by invoking the `DWI::get_valid_DW_scheme()` function with the relevant `Header` as argument. 
  - We could provide an equivalent method / function to allow the application to process the gradient import command-line options and load and replace the DW scheme in the relevant `Header`. But this introduces some scope for incorrect operation:
    - The application developer must remember to call the relevant function / method after the `Image`/`Header` `open()` call to allow the options to be processed. This could easily be overlooked since we tend to work with DICOM or `mif` images, which don't need these options. 
    - The application developer must remember to call this function _before_ copying the `Header` to prepare the `Header` for the output image(s). Otherwise the output `Header` may end up with no DW scheme, or with the unprocessed DW scheme. 

--- 

Putting all this together, I propose (thoughts welcome):

- DW scheme stored as Eigen::MatrixXD member of `Header` class. 
- provide `DW_scheme()`, `clear_DW_scheme()`, and `set_DW_scheme()` methods, with the latter performing all the necessary checks and balances. 
- add `with_DW_scheme()` method to `Header` class (along with similar wrapper in `Image`) to process command-line options. The intention there is to encourage usage of the form:
  ```
  auto in = Image<float>::open (argument[0]).with_direct_io (3).with_dw_scheme();
  ```
  which would provide all the required guarantees. 
- It's not clear to me how to handle applications like `dwicat` that intrinsically require multiple input DWI datasets. It strikes me that the only sensible thing to do here is to _not_ offer the gradient import command-line options (which is the case currently). But it is a very rare occurrence, so can probably be safely ignored. 
- With these changes, we would lose the ability to check the original unmodified version of the DW scheme - but it's very rare to need access to it anyway, so I don't think that's a great loss. And in any case, users should be able to inspect the original source using other means. 
- We also lose the ability to see what the DW scheme was if it doesn't match the data (in terms of number of volumes). Realistically, this would only affect `mrinfo`. Again, not a show-stopper. 

Optional:
- relegate the `-bvalue_scaling` option to an environment variable. It's rarely used, but takes up real estate on the help pages. We can provide access to the same functionality with an environment variable, e.g.:
  ```
  BVALUE_SCALE=0 mrconvert DICOM/ dwi.mif
  ```

**Comment by Lestropie** [comment]
We definitely put a lot of thought and debate at the time into shifting `dw_scheme` from a dedicated `Header` entry into a key-value string. That was a long time before supporting the `-bvalue_scaling` option. I'm not sure why we'd want to shift back:

- It would mean that DW scheme handling would explicitly be different to that of e.g. phase encoding information, slice encoding / timing information, potentially arbitrary 4D sidecar schemes (your suggestion from memory), ...

- Any class providing the requisite functions to be accepted as an input expecting `HeaderType` would need to provide DW scheme set / get functions even if they have nothing to do with DWI.

> But if we want to ensure that the DW scheme is always valid as soon as the image has been read ...

Think we decided at the time that we *didn't* want this to be the case. If a command doesn't need to read diffusion encoding information, then there's no reason why it should need to enforce a valid DW scheme. And for any command that does require a valid DW scheme, `get_DW_scheme()` is likely to be the next line of code anyway.

-----

Personally I've been happy with shifting the DW scheme to `keyval()`. I'd rather just:

- Relegate the `-bvalue_scaling` option to `mrconvert` only;

- Perform rescaling upon conversion to numerical form only if the maximal deviation from direction unit norm exceeds some threshold (should only happen once).

That's a far less drastic change from what we've been running with for many years, and less likely to introduce unwanted behaviour right before release.

**Comment by Lestropie** [comment]
Closed by #2015.
